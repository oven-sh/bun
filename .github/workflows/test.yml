name: Test

concurrency:
  group: ${{ github.workflow }}-${{ inputs.run-id || github.run_id }}
  cancel-in-progress: true

on:
  push:
  workflow_call:
    inputs:
      run-id:
        type: string
        required: true

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/download-artifact@v4
        with:
          run-id: ${{ inputs.run-id || '8623134412' }}
          pattern: bun-*
      - run: |
          ls -la
