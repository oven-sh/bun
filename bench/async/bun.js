import { bench, run } from "mitata";

bench("sync", () => {});
bench("async", async () => {});
bench("await 1", async () => await 1);

await run();
