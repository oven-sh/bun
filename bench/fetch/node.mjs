import { bench, run } from "mitata";

const count = 100;

bench(`fetch(https://example.com) x ${count}`, async () => {
  const requests = new Array(count);

  for (let i = 0; i < requests.length; i++) {
    requests[i] = fetch(`https://www.example.com/?cachebust=${i}`).then((r) =>
      r.text()
    );
  }

  await Promise.all(requests);
});

await run();
