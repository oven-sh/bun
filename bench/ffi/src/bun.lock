{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@napi-rs/cli": "^2.10.1",
        "@node-rs/helper": "^1.3.3",
      },
    },
  },
  "packages": {
    "@napi-rs/cli": ["@napi-rs/cli@2.14.8", "", { "bin": { "napi": "scripts/index.js" } }, "sha512-IvA3s8BqohMdUbOkFn7+23u1dhIJZCkA8Xps7DD4SLdCMbcbUF6MUuKiqxuqmVHBFTaxU25sU56WdX3efqGgPw=="],

    "@napi-rs/triples": ["@napi-rs/triples@1.1.0", "", {}, "sha512-XQr74QaLeMiqhStEhLn1im9EOMnkypp7MZOwQhGzqp2Weu5eQJbpPxWxixxlYRKWPOmJjsk6qYfYH9kq43yc2w=="],

    "@node-rs/helper": ["@node-rs/helper@1.3.3", "", { "dependencies": { "@napi-rs/triples": "^1.1.0" } }, "sha512-p4OdfQObGN9YFy5WZaGwlPYICQSe7xZYyXB0sxREmvj1HzGKp5bPg2PlfgfMZEfnjIA882B9ZrnagYzZihIwjA=="],
  }
}
