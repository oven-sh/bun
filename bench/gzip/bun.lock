{
  "lockfileVersion": 0,
  "workspaces": {
    "": {
      "name": "bench",
      "dependencies": {
        "@babel/standalone": "7.24.10",
      },
    },
  },
  "packages": {
    "@babel/standalone": ["@babel/standalone@7.24.10", "", {}, "sha512-nGC37EKfmelpyCXto1pw6SBkD5ZQRdMbL6WISi28xWit9dtiy9dChU1WgEfzturUTxrmOGkMDRrCydFMA7uOaQ=="],
  }
}
