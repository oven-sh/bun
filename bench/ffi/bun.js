import { ptr, dlopen, CString, toBuffer } from "bun:ffi";
import { run, bench, group } from "mitata";

const { napiNoop, napiHash, napiString } = require(import.meta.dir + "/src/ffi_napi_bench.node");

const {
  symbols: {
    ffi_noop: { native: ffi_noop },
    ffi_hash: { native: ffi_hash },
    ffi_string: { native: ffi_string },
  },
} = dlopen(import.meta.dir + "/src/ffi_napi_bench.node", {
  ffi_noop: { args: [], returns: "void" },
  ffi_string: { args: [], returns: "ptr" },
  ffi_hash: { args: ["ptr", "u32"], returns: "u32" },
});

const bytes = new Uint8Array(64);

group("bun:ffi", () => {
  bench("noop", () => ffi_noop());
  bench("hash", () => ffi_hash(ptr(bytes), bytes.byteLength));

  bench("c string", () => new CString(ffi_string()));
});

if (process.env.SHOW_NAPI)
  group("bun:napi", () => {
    bench("noop", () => napiNoop());
    bench("hash", () => napiHash(bytes));

    bench("string", () => napiString());
  });

await run();
