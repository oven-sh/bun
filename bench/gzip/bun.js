import { run, bench, group } from "mitata";
import { gzipSync, gunzipSync } from "bun";

const data = new TextEncoder().encode("Hello World!".repeat(9999));

const compressed = gzipSync(data);

const libraries = ["zlib"];
if (Bun.semver.satisfies(Bun.version, ">=1.1.21")) {
  libraries.push("libdeflate");
}

const options = { library: undefined };
const benchFn = (name, fn) => {
  if (libraries.length > 1) {
    group(name, () => {
      for (const library of libraries) {
        bench(library, () => {
          options.library = library;
          fn();
        });
      }
    });
  } else {
    bench(name, () => {
      fn();
    });
  }
};

benchFn(`roundtrip - "Hello World!".repeat(9999))`, () => {
  gunzipSync(gzipSync(data, options), options);
});

benchFn(`gzipSync("Hello World!".repeat(9999)))`, () => {
  gzipSync(data, options);
});

benchFn(`gunzipSync("Hello World!".repeat(9999)))`, () => {
  gunzipSync(compressed, options);
});

await run();
