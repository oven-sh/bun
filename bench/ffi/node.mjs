import { bench, group, run } from "mitata";
import { createRequire } from "node:module";

const require = createRequire(import.meta.url);
const { napiNoop, napiHash, napiString } = require("./src/ffi_napi_bench.node");

const bytes = new Uint8Array(64);

group("napi", () => {
  bench("noop", () => napiNoop());
  bench("hash", () => napiHash(bytes));

  bench("string", () => napiString());
});

await run();
