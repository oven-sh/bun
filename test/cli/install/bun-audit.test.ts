import { readableStreamToText, spawn } from "bun";
import { afterAll, beforeAll, describe, expect, test } from "bun:test";
import { bunEnv, bunExe, DirectoryTree, gunzipJsonRequest, lazyPromiseLike, tempDirWithFiles } from "harness";
import { join } from "node:path";
import { resolveBulkAdvisoryFixture } from "./registry/fixtures/audit/audit-fixtures";

function fixture(
  folder:
    | "express@3"
    | "vuln-with-only-dev-dependencies"
    | "safe-is-number@7"
    | "mix-of-safe-and-vulnerable-dependencies",
) {
  return join(import.meta.dirname, "registry", "fixtures", "audit", folder);
}

let server: Bun.Server;

beforeAll(() => {
  server = Bun.serve({
    fetch: async req => {
      const body = await gunzipJsonRequest(req);

      const fixture = resolveBulkAdvisoryFixture(body);

      if (!fixture) {
        console.log("No fixture found for", body);
        return new Response("No fixture found", { status: 404 });
      }

      return Response.json(fixture);
    },
  });
});

afterAll(() => {
  server.stop();
});

function doAuditTest(
  label: string,
  options: {
    args?: string[];
    exitCode: number;
    files: DirectoryTree | string;
    fn: (std: { stdout: PromiseLike<string>; stderr: PromiseLike<string>; dir: string }) => Promise<void>;
  },
) {
  test(label, async () => {
    const dir = tempDirWithFiles("bun-test-pm-audit-" + label.replace(/[^a-zA-Z0-9]/g, "-"), options.files);

    const cmd = [bunExe(), "pm", "audit", ...(options.args ?? [])];

    const url = server.url.toString().slice(0, -1);

    const proc = spawn({
      cmd,
      stdout: "pipe",
      stderr: "pipe",
      cwd: dir,
      env: {
        ...bunEnv,
        NPM_CONFIG_REGISTRY: url,
      },
    });

    const stdout = lazyPromiseLike(() => readableStreamToText(proc.stdout));
    const stderr = lazyPromiseLike(() => readableStreamToText(proc.stderr));

    const exitCode = await proc.exited;

    try {
      expect(exitCode).toBe(options.exitCode);
      await options.fn({ stdout, stderr, dir });
    } catch (e) {
      const err = await stderr;
      const out = await stdout;

      // useful to see what went wrong otherwise
      // we are just eating the rror silently

      console.log("ERR:", err);
      console.log("OUT:", out);

      throw e; //but still rethrow so test fails
    }
  });
}

describe("`bun pm audit`", () => {
  doAuditTest("should fail with no package.json", {
    exitCode: 1,
    files: {
      "README.md": "This place sure is empty...",
    },
    fn: async ({ stderr }) => {
      expect(await stderr).toContain("No package.json was found for directory");
    },
  });

  doAuditTest("should fail with package.json but no lockfile", {
    exitCode: 1,
    files: {
      "package.json": JSON.stringify({
        name: "test",
        version: "1.0.0",
        dependencies: {
          "express": "3",
        },
      }),
    },
    fn: async ({ stderr }) => {
      expect(await stderr).toContain("error: Lockfile not found");
    },
  });

  doAuditTest("should exit 0 when there are no dependencies in package.json", {
    exitCode: 0,
    files: {
      // i deemed this small enough to justify not needing a fixture
      "package.json": JSON.stringify({
        name: "empty-package",
        version: "1.0.0",
      }),
      "bun.lock": JSON.stringify({
        "lockfileVersion": 1,
        "workspaces": {
          "": {
            "name": "empty-package",
          },
        },
        "packages": {},
      }),
    },
    fn: async ({ stdout }) => {
      expect(await stdout).toBe("No vulnerabilities found\n");
    },
  });

  doAuditTest("should exit 0 when there are no vulnerabilities", {
    exitCode: 0,
    files: fixture("safe-is-number@7"),
    fn: async ({ stdout }) => {
      expect(await stdout).toBe("No vulnerabilities found\n");
    },
  });

  doAuditTest("should exit code 1 when there are vulnerabilities", {
    exitCode: 1,
    files: fixture("express@3"),
    fn: async ({ stdout }) => {
      expect(await stdout).toMatchSnapshot("bun-audit-expect-vulnerabilities-found");
    },
  });

  doAuditTest("should print valid JSON and exit 0 when --json is passed and there are no vulnerabilities", {
    exitCode: 0,
    files: fixture("safe-is-number@7"),
    args: ["--json"],
    fn: async ({ stdout }) => {
      const out = await stdout;
      const json = JSON.parse(out); // this would throw making the test fail if the JSON was invalid
      expect(json).toMatchSnapshot("bun-audit-expect-valid-json-stdout-report-no-vulnerabilities");
    },
  });

  doAuditTest("should print valid JSON and exit 0 when --json is passed and there are vulnerabilities", {
    exitCode: 0,
    files: fixture("express@3"),
    args: ["--json"],
    fn: async ({ stdout }) => {
      const out = await stdout;
      const json = JSON.parse(out); // this would throw making the test fail if the JSON was invalid
      expect(json).toMatchSnapshot("bun-audit-expect-valid-json-stdout-report-vulnerabilities");
    },
  });

  doAuditTest(
    "should exit 1 and behave exactly the same when there are vulnerabilities when only devDependencies are specified",
    {
      exitCode: 1,
      files: fixture("vuln-with-only-dev-dependencies"),
      fn: async ({ stdout }) => {
        expect(await stdout).toMatchSnapshot("bun-audit-expect-vulnerabilities-found");
      },
    },
  );

  doAuditTest(
    "when a project has some safe dependencies and some vulnerable dependencies, we should not print the safe dependencies",
    {
      exitCode: 1,
      files: fixture("mix-of-safe-and-vulnerable-dependencies"),
      fn: async ({ stdout }) => {
        // this fixture is using a safe version of is-number and an unsafe version of ms
        // so we want to check that `is-number` is not included in the output and that `ms` is

        const out = await stdout;

        expect(out).toContain("ms");
        expect(out).not.toContain("is-number");

        expect(out).toMatchSnapshot("bun-audit-expect-vulnerabilities-found");
      },
    },
  );

  const fakeIntegrity = // this is just random/fake data as the integrity check is not important for this test
    "sha512-V8E0l1jyyeSSS9R+J9oljx5eq2rqzClInuwaPcyuv0Mm3ViI/3/rcc4rCEO8i4eQ4I0O0FAGYDA2i5xWHHPhzg==";

  doAuditTest(
    "packages that come from non-default registries should be ignored from the audit, however they should get surfaced at the bottom of the output that they got skipped",
    {
      exitCode: 0,
      files: {
        "package.json": JSON.stringify({
          name: "test",
          version: "1.0.0",
          dependencies: {
            "@foo/bar": "1.0.0",
            "@foo/baz": "1.0.0",
          },
        }),
        "bun.lock": JSON.stringify({
          "lockfileVersion": 1,
          "workspaces": {
            "": {
              "name": "test",
            },
          },
          "packages": {
            "@foo/bar": ["@foo/bar@1.0.0", "", {}, fakeIntegrity],
            "@foo/baz": ["@foo/baz@1.0.0", "", {}, fakeIntegrity],
          },
        }),
        //prettier-ignore
        ".npmrc": [
          `registry=https://registry.npmjs.org`,
          `@foo:registry=https://my-registry.example.com`,
        ].join("\n"),
      },
      fn: async ({ stdout }) => {
        const out = await stdout;

        expect(out).toContain("Skipped @foo/bar, @foo/baz because they do not come from the default registry");
        expect(out).toContain("No vulnerabilities found");
      },
    },
  );

  doAuditTest("workspaces print the path to the vulnerable package and include workspace:pkg in the name", {
    exitCode: 1,
    files: {
      "package.json": JSON.stringify({
        name: "test",
        version: "1.0.0",
        workspaces: ["a"],
      }),

      "a/package.json": JSON.stringify({
        "name": "a",
        "dependencies": {
          "ms": "0.7.0",
        },
      }),

      "bun.lock": JSON.stringify({
        "lockfileVersion": 1,
        "workspaces": {
          "": {
            "name": "bun-audit-playground",
          },
          "a": {
            "name": "a",
            "dependencies": {
              "ms": "0.7.0",
            },
          },
        },
        "packages": {
          "a": ["a@workspace:a"],
          "ms": ["ms@0.7.0", "", {}, fakeIntegrity],
        },
      }),
    },
    fn: async ({ stdout }) => {
      expect(await stdout).toInclude("workspace:a â€º ms");
    },
  });
});
