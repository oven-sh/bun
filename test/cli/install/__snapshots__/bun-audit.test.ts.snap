// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`\`bun pm audit\` should exit code 1 when there are vulnerabilities: bun-audit-expect-vulnerabilities-found 1`] = `
"minimist  <0.2.4
  express › mkdirp › minimist
  critical: Prototype Pollution in minimist - https://github.com/advisories/GHSA-xvch-5gv4-984h
  moderate: Prototype Pollution in minimist - https://github.com/advisories/GHSA-vh95-rmgr-6w4m

express  >=3.4.5 <4.0.0-rc1
  (direct dependency)
  low: Express Open Redirect vulnerability - https://github.com/advisories/GHSA-jj78-5fmv-mv28
  low: express vulnerable to XSS via response.redirect() - https://github.com/advisories/GHSA-qw6h-vgh9-j6wx
  moderate: Express ressource injection - https://github.com/advisories/GHSA-cm5g-3pgc-8rg4
  moderate: Express.js Open Redirect in malformed URLs - https://github.com/advisories/GHSA-rv95-896h-c2vc

qs  <6.2.4
  express › connect › body-parser › qs
  high: qs vulnerable to Prototype Pollution - https://github.com/advisories/GHSA-hrpp-h998-j3pp
  high: Prototype Pollution Protection Bypass in qs - https://github.com/advisories/GHSA-gqgv-6jq5-jjj9

send  <0.19.0
  express › send
  low: send vulnerable to template injection that can lead to XSS - https://github.com/advisories/GHSA-m6fv-jmcg-4jfg

negotiator  <0.6.1
  express › connect › serve-index › accepts › negotiator
  high: Regular Expression Denial of Service in negotiator - https://github.com/advisories/GHSA-7mc5-chhp-fmc3

base64-url  <2.0.0
  express › connect › csurf › csrf › uid-safe › base64-url
  high: Out-of-bounds Read in base64-url - https://github.com/advisories/GHSA-j4mr-9xw3-c9jx

serve-static  <1.16.0
  express › connect › serve-static
  low: serve-static vulnerable to template injection that can lead to XSS - https://github.com/advisories/GHSA-cm22-4g7w-348p

cookie  <0.7.0
  express › connect › cookie
  low: cookie accepts cookie name, path, and domain with out of bounds characters - https://github.com/advisories/GHSA-pxg6-pf52-xh8x

mime  <1.4.1
  express › send › mime
  high: mime Regular Expression Denial of Service when MIME lookup performed on untrusted user input - https://github.com/advisories/GHSA-wrvr-8mpx-r7pp

body-parser  <1.20.3
  express › connect › body-parser
  high: body-parser vulnerable to denial of service when url encoding is enabled - https://github.com/advisories/GHSA-qwcr-r2fm-qrc7

fresh  <0.5.2
  express › connect › fresh
  high: Regular Expression Denial of Service in fresh - https://github.com/advisories/GHSA-9qj9-36jm-prpv

morgan  <1.9.1
  express › connect › morgan
  critical: Code Injection in morgan - https://github.com/advisories/GHSA-gwg9-rgvj-4h5j

basic-auth-connect  <1.1.0
  express › connect › basic-auth-connect
  high: basic-auth-connect's callback uses time unsafe string comparison - https://github.com/advisories/GHSA-7p89-p6hx-q4fw

debug  <2.6.9
  express › connect › compression › debug
  high: debug Inefficient Regular Expression Complexity vulnerability - https://github.com/advisories/GHSA-9vvw-cc9w-f27h
  low: Regular Expression Denial of Service in debug - https://github.com/advisories/GHSA-gxpj-cx7g-858c

ms  <2.0.0
  express › connect › serve-favicon › ms
  moderate: Vercel ms Inefficient Regular Expression Complexity vulnerability - https://github.com/advisories/GHSA-w9mr-4mfr-499f

21 vulnerabilities (2 critical, 9 high, 4 moderate, 6 low)

To update all dependencies to the latest compatible versions:
  bun update

To update all dependencies to the latest versions (including breaking changes):
  bun update --latest

"
`;

exports[`\`bun pm audit\` should print valid JSON and exit 0 when --json is passed and there are no vulnerabilities: bun-audit-expect-valid-json-stdout-report-no-vulnerabilities 1`] = `{}`;

exports[`\`bun pm audit\` should print valid JSON and exit 0 when --json is passed and there are vulnerabilities: bun-audit-expect-valid-json-stdout-report-vulnerabilities 1`] = `
{
  "base64-url": [
    {
      "cvss": {
        "score": 0,
        "vectorString": null,
      },
      "cwe": [
        "CWE-125",
      ],
      "id": 1090859,
      "severity": "high",
      "title": "Out-of-bounds Read in base64-url",
      "url": "https://github.com/advisories/GHSA-j4mr-9xw3-c9jx",
      "vulnerable_versions": "<2.0.0",
    },
  ],
  "basic-auth-connect": [
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:N/A:N",
      },
      "cwe": [
        "CWE-208",
      ],
      "id": 1099800,
      "severity": "high",
      "title": "basic-auth-connect's callback uses time unsafe string comparison",
      "url": "https://github.com/advisories/GHSA-7p89-p6hx-q4fw",
      "vulnerable_versions": "<1.1.0",
    },
  ],
  "body-parser": [
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
      },
      "cwe": [
        "CWE-405",
      ],
      "id": 1099520,
      "severity": "high",
      "title": "body-parser vulnerable to denial of service when url encoding is enabled",
      "url": "https://github.com/advisories/GHSA-qwcr-r2fm-qrc7",
      "vulnerable_versions": "<1.20.3",
    },
  ],
  "cookie": [
    {
      "cvss": {
        "score": 0,
        "vectorString": null,
      },
      "cwe": [
        "CWE-74",
      ],
      "id": 1103907,
      "severity": "low",
      "title": "cookie accepts cookie name, path, and domain with out of bounds characters",
      "url": "https://github.com/advisories/GHSA-pxg6-pf52-xh8x",
      "vulnerable_versions": "<0.7.0",
    },
  ],
  "debug": [
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
      },
      "cwe": [
        "CWE-1333",
      ],
      "id": 1094457,
      "severity": "high",
      "title": "debug Inefficient Regular Expression Complexity vulnerability",
      "url": "https://github.com/advisories/GHSA-9vvw-cc9w-f27h",
      "vulnerable_versions": "<2.6.9",
    },
    {
      "cvss": {
        "score": 3.7,
        "vectorString": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:N/A:L",
      },
      "cwe": [
        "CWE-400",
      ],
      "id": 1096795,
      "severity": "low",
      "title": "Regular Expression Denial of Service in debug",
      "url": "https://github.com/advisories/GHSA-gxpj-cx7g-858c",
      "vulnerable_versions": "<2.6.9",
    },
  ],
  "express": [
    {
      "cvss": {
        "score": 4.7,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:N/A:N",
      },
      "cwe": [
        "CWE-601",
      ],
      "id": 1099969,
      "severity": "low",
      "title": "Express Open Redirect vulnerability",
      "url": "https://github.com/advisories/GHSA-jj78-5fmv-mv28",
      "vulnerable_versions": ">=3.4.5 <4.0.0-rc1",
    },
    {
      "cvss": {
        "score": 5,
        "vectorString": "CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:U/C:L/I:L/A:L",
      },
      "cwe": [
        "CWE-79",
      ],
      "id": 1100530,
      "severity": "low",
      "title": "express vulnerable to XSS via response.redirect()",
      "url": "https://github.com/advisories/GHSA-qw6h-vgh9-j6wx",
      "vulnerable_versions": "<4.20.0",
    },
    {
      "cvss": {
        "score": 4,
        "vectorString": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:C/C:L/I:N/A:N",
      },
      "cwe": [
        "CWE-74",
      ],
      "id": 1101381,
      "severity": "moderate",
      "title": "Express ressource injection",
      "url": "https://github.com/advisories/GHSA-cm5g-3pgc-8rg4",
      "vulnerable_versions": "<=3.21.4",
    },
    {
      "cvss": {
        "score": 6.1,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N",
      },
      "cwe": [
        "CWE-601",
        "CWE-1286",
      ],
      "id": 1096820,
      "severity": "moderate",
      "title": "Express.js Open Redirect in malformed URLs",
      "url": "https://github.com/advisories/GHSA-rv95-896h-c2vc",
      "vulnerable_versions": "<4.19.2",
    },
  ],
  "fresh": [
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
      },
      "cwe": [
        "CWE-400",
      ],
      "id": 1093570,
      "severity": "high",
      "title": "Regular Expression Denial of Service in fresh",
      "url": "https://github.com/advisories/GHSA-9qj9-36jm-prpv",
      "vulnerable_versions": "<0.5.2",
    },
  ],
  "mime": [
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
      },
      "cwe": [
        "CWE-400",
      ],
      "id": 1093780,
      "severity": "high",
      "title": "mime Regular Expression Denial of Service when MIME lookup performed on untrusted user input",
      "url": "https://github.com/advisories/GHSA-wrvr-8mpx-r7pp",
      "vulnerable_versions": "<1.4.1",
    },
  ],
  "minimist": [
    {
      "cvss": {
        "score": 9.8,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
      },
      "cwe": [
        "CWE-1321",
      ],
      "id": 1097677,
      "severity": "critical",
      "title": "Prototype Pollution in minimist",
      "url": "https://github.com/advisories/GHSA-xvch-5gv4-984h",
      "vulnerable_versions": "<0.2.4",
    },
    {
      "cvss": {
        "score": 5.6,
        "vectorString": "CVSS:3.1/AV:N/AC:H/PR:N/UI:N/S:U/C:L/I:L/A:L",
      },
      "cwe": [
        "CWE-1321",
      ],
      "id": 1096466,
      "severity": "moderate",
      "title": "Prototype Pollution in minimist",
      "url": "https://github.com/advisories/GHSA-vh95-rmgr-6w4m",
      "vulnerable_versions": "<0.2.1",
    },
  ],
  "morgan": [
    {
      "cvss": {
        "score": 9.8,
        "vectorString": "CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
      },
      "cwe": [
        "CWE-94",
      ],
      "id": 1093804,
      "severity": "critical",
      "title": "Code Injection in morgan",
      "url": "https://github.com/advisories/GHSA-gwg9-rgvj-4h5j",
      "vulnerable_versions": "<1.9.1",
    },
  ],
  "ms": [
    {
      "cvss": {
        "score": 5.3,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:L",
      },
      "cwe": [
        "CWE-1333",
      ],
      "id": 1094419,
      "severity": "moderate",
      "title": "Vercel ms Inefficient Regular Expression Complexity vulnerability",
      "url": "https://github.com/advisories/GHSA-w9mr-4mfr-499f",
      "vulnerable_versions": "<2.0.0",
    },
  ],
  "negotiator": [
    {
      "cvss": {
        "score": 0,
        "vectorString": null,
      },
      "cwe": [
        "CWE-400",
      ],
      "id": 1090969,
      "severity": "high",
      "title": "Regular Expression Denial of Service in negotiator",
      "url": "https://github.com/advisories/GHSA-7mc5-chhp-fmc3",
      "vulnerable_versions": "<0.6.1",
    },
  ],
  "qs": [
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
      },
      "cwe": [
        "CWE-1321",
      ],
      "id": 1104115,
      "severity": "high",
      "title": "qs vulnerable to Prototype Pollution",
      "url": "https://github.com/advisories/GHSA-hrpp-h998-j3pp",
      "vulnerable_versions": "<6.2.4",
    },
    {
      "cvss": {
        "score": 7.5,
        "vectorString": "CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
      },
      "cwe": [
        "CWE-20",
      ],
      "id": 1087527,
      "severity": "high",
      "title": "Prototype Pollution Protection Bypass in qs",
      "url": "https://github.com/advisories/GHSA-gqgv-6jq5-jjj9",
      "vulnerable_versions": "<6.0.4",
    },
  ],
  "send": [
    {
      "cvss": {
        "score": 5,
        "vectorString": "CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:U/C:L/I:L/A:L",
      },
      "cwe": [
        "CWE-79",
      ],
      "id": 1100526,
      "severity": "low",
      "title": "send vulnerable to template injection that can lead to XSS",
      "url": "https://github.com/advisories/GHSA-m6fv-jmcg-4jfg",
      "vulnerable_versions": "<0.19.0",
    },
  ],
  "serve-static": [
    {
      "cvss": {
        "score": 5,
        "vectorString": "CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:U/C:L/I:L/A:L",
      },
      "cwe": [
        "CWE-79",
      ],
      "id": 1100528,
      "severity": "low",
      "title": "serve-static vulnerable to template injection that can lead to XSS",
      "url": "https://github.com/advisories/GHSA-cm22-4g7w-348p",
      "vulnerable_versions": "<1.16.0",
    },
  ],
}
`;

exports[`\`bun pm audit\` should exit 1 and behave exactly the same when there are vulnerabilities when only devDependencies are specified: bun-audit-expect-vulnerabilities-found 1`] = `
"ms  <2.0.0
  (direct dependency)
  moderate: Vercel ms Inefficient Regular Expression Complexity vulnerability - https://github.com/advisories/GHSA-w9mr-4mfr-499f
  high: Regular Expression Denial of Service in ms - https://github.com/advisories/GHSA-3fx5-fwvr-xrjg

2 vulnerabilities (1 high, 1 moderate)

To update all dependencies to the latest compatible versions:
  bun update

To update all dependencies to the latest versions (including breaking changes):
  bun update --latest

"
`;

exports[`\`bun pm audit\` when a project has some safe dependencies and some vulnerable dependencies, we should not print the safe dependencies: bun-audit-expect-vulnerabilities-found 1`] = `
"ms  <2.0.0
  (direct dependency)
  moderate: Vercel ms Inefficient Regular Expression Complexity vulnerability - https://github.com/advisories/GHSA-w9mr-4mfr-499f
  high: Regular Expression Denial of Service in ms - https://github.com/advisories/GHSA-3fx5-fwvr-xrjg

2 vulnerabilities (1 high, 1 moderate)

To update all dependencies to the latest compatible versions:
  bun update

To update all dependencies to the latest versions (including breaking changes):
  bun update --latest

"
`;
